File: py/graph/dijkstra.py
from typing import List, Dict
from math import inf
from weighted import plain_graph, A, D, B, C, Y, WeightedGraph
from heapq import heappop, heappush


def find_lowest_cost(nodes: Dict[str, int or None]) -> Dict[str, int] or None:
    lowest = inf
    node = None

    for name, cost in nodes.items():
        if cost < lowest:
            lowest = cost
            node = name

    return node


def dijkstra(graph: WeightedGraph, start: str, end: str) -> (List[str], int) or None:
    parents = {start: None}
    costs = {node: inf for node in graph}
    costs[start] = 0

    priority_queue = [(0, start)]
    visited = set()

    while priority_queue:
        cost, node = heappop(priority_queue)
        neighbors = graph[node]

        if node == end:
            path = []
            parent = node

            while parent:
                path.append(parent)
                parent = parents[parent]

            return path[::-1], cost

        if node in visited:
            continue
        visited.add(node)

        for neighbor, weight in neighbors.items():
            new_cost = cost + weight
            if new_cost < costs[neighbor]:
                costs[neighbor] = new_cost
                parents[neighbor] = node
                heappush(priority_queue, (new_cost, neighbor))

    return None


print(dijkstra(plain_graph, A, B))

File: ts/src/graph/dijkstra.ts
import { PHeap } from '../helpers/PHeap'
import { A, D, WeightedGraph, plainGraph } from './weighted'

const dijkstra = (graph: WeightedGraph, start: string, end: string): { path: string[]; cost: number } | null => {
	const costs = Object.fromEntries(Object.keys(graph).map((node) => [node, Infinity]))
	costs[start] = 0
	const parents: Record<string, string | null> = { [start]: null }
	const visited = new Set()

	const heap = new PHeap()
	heap.push([0, start])

	while (heap.heap.length > 0) {
		const [cost, current] = heap.pop() as [number, string]
		const neighbors = graph[current]

		if (current === end) {
			let parent = current
			const path = []

			while (parent) {
				path.push(parent)
				parent = parents[parent] as string
			}

			return { path: path.reverse(), cost }
		}

		if (visited.has(current)) continue
		visited.add(current)

		for (let [neighbor, weight] of Object.entries(neighbors)) {
			const newCost = weight + cost

			if (costs[neighbor] > newCost) {
				costs[neighbor] = newCost
				parents[neighbor] = current
				heap.push([newCost, neighbor])
			}
		}
	}

	return null
}

console.log(dijkstra(plainGraph, A, D))
